From fca09160bc84c101d921fab8b45e6b2f0a4bb14d Mon Sep 17 00:00:00 2001
From: Carlos Calderon <carlos@emuex.com>
Date: Thu, 13 Sep 2018 16:23:45 +0100
Subject: [PATCH 14/19] upboard: Add support for UP-CRST02

UP-CRST02 uses
* ACPI ID AANT0F03
* same pin mappings as UP-CREX
* same register ranges as UP2

Add support to MFD and pinctrl.
---
 drivers/mfd/upboard-fpga.c        | 23 +++++++++++++++++++----
 drivers/pinctrl/pinctrl-upboard.c | 19 +++++++++++++++++++
 2 files changed, 38 insertions(+), 4 deletions(-)

diff --git a/drivers/mfd/upboard-fpga.c b/drivers/mfd/upboard-fpga.c
index 3ce27e2..0e81929 100644
--- a/drivers/mfd/upboard-fpga.c
+++ b/drivers/mfd/upboard-fpga.c
@@ -149,6 +149,20 @@ static const struct upboard_fpga_data upboard_up2_fpga_data = {
 	.ncells = ARRAY_SIZE(upboard_up2_mfd_cells),
 };
 
+/* UP-CREX carrier board for UP Core */
+
+/* same MAXV config as UP1 (proto2 release) */
+#define upboard_upcore_crex_fpga_data upboard_up_fpga_data
+
+/* UP-CRST02 carrier board for UP Core */
+
+/* same MAX10 config as UP2, but same LED cells as UP1 */
+static const struct upboard_fpga_data upboard_upcore_crst02_fpga_data = {
+	.regmapconf = &upboard_up2_regmap_config,
+	.cells = upboard_up_mfd_cells,
+	.ncells = ARRAY_SIZE(upboard_up_mfd_cells),
+};
+
 static int upboard_fpga_read(void *context, unsigned int reg, unsigned int *val)
 {
 	struct upboard_fpga * const fpga = context;
@@ -290,10 +304,11 @@ static int __init upboard_fpga_detect_firmware(struct upboard_fpga *fpga)
 }
 
 static const struct acpi_device_id upboard_fpga_acpi_match[] = {
-	{ "AANT0F00", (kernel_ulong_t) &upboard_up_fpga_data }, /*UP1*/
-	{ "AANT0F01", (kernel_ulong_t) &upboard_up2_fpga_data },/*UP2*/
-	{ "AANT0F02", (kernel_ulong_t) &upboard_up_fpga_data }, /*UPCORE CREX*/
-	{ },
+	{ "AANT0F00", (kernel_ulong_t)&upboard_up_fpga_data },
+	{ "AANT0F01", (kernel_ulong_t)&upboard_up2_fpga_data },
+	{ "AANT0F02", (kernel_ulong_t)&upboard_upcore_crex_fpga_data },
+	{ "AANT0F03", (kernel_ulong_t)&upboard_upcore_crst02_fpga_data },
+	{ }
 };
 MODULE_DEVICE_TABLE(acpi, upboard_fpga_acpi_match);
 
diff --git a/drivers/pinctrl/pinctrl-upboard.c b/drivers/pinctrl/pinctrl-upboard.c
index 65ac489..ed1d9bd 100644
--- a/drivers/pinctrl/pinctrl-upboard.c
+++ b/drivers/pinctrl/pinctrl-upboard.c
@@ -535,6 +535,13 @@ static unsigned int upboard_upcore_crex_rpi_mapping[] = {
 	UPBOARD_UPCORE_CREX_BIT_TO_PIN(1, I2S2_TX),
 };
 
+/*
+ * UP Core board + CRST02 carrier board data
+ */
+
+#define upboard_upcore_crst02_pins        upboard_upcore_crex_pins
+#define upboard_upcore_crst02_rpi_mapping upboard_upcore_crex_rpi_mapping
+
 static int upboard_set_mux(struct pinctrl_dev *pctldev, unsigned int function,
 			   unsigned int group)
 {
@@ -652,6 +659,15 @@ static struct pinctrl_desc upboard_upcore_crex_pinctrl_desc = {
 	.owner = THIS_MODULE,
 };
 
+static struct pinctrl_desc upboard_upcore_crst02_pinctrl_desc = {
+	.pins = upboard_upcore_crst02_pins,
+	.npins = ARRAY_SIZE(upboard_upcore_crst02_pins),
+	.pctlops = &upboard_pinctrl_ops,
+	.pmxops = &upboard_pinmux_ops,
+	.owner = THIS_MODULE,
+};
+
+
 static int upboard_rpi_to_native_gpio(struct gpio_chip *gc, unsigned int gpio)
 {
 	struct upboard_pinctrl *pctrl =
@@ -793,6 +809,9 @@ static int __init upboard_pinctrl_probe(struct platform_device *pdev)
 	} else if (!strcmp(hid, "AANT0F02")) {
 		pctldesc = &upboard_upcore_crex_pinctrl_desc;
 		rpi_mapping = upboard_upcore_crex_rpi_mapping;
+	} else if (!strcmp(hid, "AANT0F03")) {
+		pctldesc = &upboard_upcore_crst02_pinctrl_desc;
+		rpi_mapping = upboard_upcore_crst02_rpi_mapping;
 	} else
 		return -ENODEV;
 
-- 
2.7.4

